### PR TITLE
Don't throw InvalidCastException, throw customized InvalidOperationException instead

### PR DESCRIPTION
**Pull request by AlexanderTaeschner**
Addresses #419

### CODE REVIEW

File: src/Microsoft.Data.Sqlite.Core/SqliteDataRecord.cs [hunk 1]
```diff
@@ -53,7 +54,7 @@ protected override int GetSqliteType(int ordinal)
         protected override T GetNull<T>()
             => typeof(T) == typeof(DBNull) || typeof(T) == typeof(object)
                 ? (T)(object)DBNull.Value
-                : throw new InvalidCastException();
+                : throw new SqlNullValueException();
```
Comment (bricelam) [inline src/Microsoft.Data.Sqlite.Core/SqliteDataRecord.cs]: I think this type is SQL Client specific and we should introduce our own type. I'll look more into it...

Comment (bricelam) [reply src/Microsoft.Data.Sqlite.Core/SqliteDataRecord.cs]: Confirmed.

> The System.Data.SqlTypes namespace provides classes for native data types in SQL Server.

I think we should just throw `InvlaidOperationException` with a message like this:

> The data is NULL. This method can't be called on NULL values. Check using IsDBNull before calling.

**Comment by AlexanderTaeschner** [comment]
Updated patch with customized InvalidOperationException.

File: test/Microsoft.Data.Sqlite.Tests/SqliteConnectionTest.cs [hunk 2]
```diff
@@ -538,7 +538,7 @@ public void CreateFunction_works_when_parameter_null_but_type_long()
                 var ex = Assert.Throws<SqliteException>(() => connection.ExecuteScalar<long>("SELECT test(NULL);"));
 
                 Assert.Equal(
-                    Resources.SqliteNativeError(raw.SQLITE_ERROR, new InvalidCastException().Message),
+                    Resources.SqliteNativeError(raw.SQLITE_ERROR, Resources.CalledOnNullValue),
```
Comment (bricelam) [inline test/Microsoft.Data.Sqlite.Tests/SqliteConnectionTest.cs]: Hmm, the error message doesn't make much sense in this context. Can we customize it for SqliteParameterBinder?

Comment (AlexanderTaeschner) [reply test/Microsoft.Data.Sqlite.Tests/SqliteConnectionTest.cs]: Okay, done in the newest revision.

**Comment by AlexanderTaeschner** [comment]
Rebased on top of current dev.

File: src/Microsoft.Data.Sqlite.Core/SqliteValueReader.cs [hunk 3]
```diff
@@ -117,24 +118,28 @@ public virtual int GetInt32(int ordinal)
...
-                ? throw new InvalidCastException()
+                ? throw new InvalidOperationException(Resources.CalledOnNullValue)
                 : GetStringCore(ordinal);
 
         protected abstract string GetStringCore(int ordinal);
 
         public virtual T GetFieldValue<T>(int ordinal)
+            => GetFieldValue<T>(ordinal, Resources.CalledOnNullValue);
+
+        public virtual T GetFieldValue<T>(int ordinal, string onNullErrorMsg)
```
Comment (bricelam) [inline src/Microsoft.Data.Sqlite.Core/SqliteValueReader.cs]: Callers should have to pass in the message. Can we simply override it (via an protected abstract property or something) in `SqliteDataRecord` and `SqliteParamterReader`? You can flow the function name in through the `SqliteParameterReader` constructor.

**Comment by AlexanderTaeschner** [comment]
Rebased on current dev and added protected abstract property with error message to SqliteValueReader.

**Pull request merged**